### DISCUSSION TITLE
Dose MILVUS-JAVA-SDK support cdc？

### DISCUSSION THREAD

**Discussion by pmupkin** [milvus-io/milvus#27716] [Q&A and General discussion]
Milvus currently supports the CDC function, but does the corresponding java-sdk-milvus support the CDC function?

**Comment by yhmo** [comment]
Milvus-CDC is an independent tool, not a sdk. It is written by Golang, so it integrates the go-sdk to do some special work.
Go-sdk, java sdk, node sdk, python sdk... neither of the sdk have the ability to capture change data.
We can say "cdc tool works with the help of sdk", but it doesn't mean "sdk can capture change data".

**Reply by pmupkin** [reply]
I use milvus in my java project. If I want to implement CDC function, how should I do it? Should I refer to milvus-cdc to implement the java version?

**Reply by yhmo** [reply]
The cdc tool does these things:
- watches the pulsar service of the source cluster. once an insert/delete request is sent to the source cluster, the cdc tool will detect and pass the Insert/delete request to the pulsar service of the target cluster.
- watches the etcd of the source cluster to get information(schema, index type, etc) of collections/partitions. once a createCollection/dropCollection/createPartition/createIndex/dropIndex request is sent to the source cluster, the cdc tool will pass the  CreateCollection/CreatePartition/DropCollection/DropPartition request to the target cluster by Go sdk.

No matter the tool is written by which language, it must have the ability to detect the DDL/DML/DCL requests of the source cluster and pass the requests to the target cluster.
- DDL: createDatabase/dropDatabase/createCollection/createPartition/dropCollection/dropPartition/createIndex/dropIndex/createAlias/dropAlias/alterAlias/
- DML: insert/delete/upsert/import
- DCL: createRole/deleteRole/createUser/dropUser/grant


Yes, you have to implement it in Java. You can read the source code of cdc tool to get more info. The tool is still in development, some tasks are not finished such as maintaining the load/release state of collections to be consistent between the source cluster and target cluster.
